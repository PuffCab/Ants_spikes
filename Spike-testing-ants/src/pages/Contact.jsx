import React, { useEffect, useState } from "react";

function Contact() {
  const [name, setName] = useState("");
  const [email, setEmail] = useState("");
  const [isVisible, setIsVisible] = useState(false);
  const [userName, setUserName] = useState(null);

  //   const getContactMessage = () => {
  //     setIsVisible(true);
  //   };
  const getContactMessageAsync = () => {
    setTimeout(() => {
      setIsVisible(true);
    }, 2000);
  };

  const fetchUser = async () => {
    const response = await fetch(
      "https://jsonplaceholder.typicode.com/users/2"
    );
    const result = await response.json();
    console.log("result :>> ", result);
    setUserName(result.username);
  };

  useEffect(() => {
    fetchUser();
  }, []);

  return (
    <div>
      <h1>Contact</h1>
      <span
        style={{ visibility: isVisible ? "visible" : "hidden" }}
        data-testid="test-span"
      >
        Thank you {name}, we will send an email to {email}
      </span>
      <div>
        <input
          type="text"
          name="name"
          id="name"
          placeholder="Name"
          value={name}
          onChange={(e) => setName(e.target.value)}
        />
        <input
          type="text"
          name="email"
          id="email"
          placeholder="Email"
          value={email}
          onChange={(e) => setEmail(e.target.value)}
        />
      </div>
      <button data-testid="test-button" onClick={getContactMessageAsync}>
        Contact me
      </button>
      {userName && (
        <div>
          <h3>userName: {userName}</h3>
        </div>
      )}
    </div>
  );
}

export default Contact;
