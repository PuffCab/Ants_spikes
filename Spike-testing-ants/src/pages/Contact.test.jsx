import { expect, test, vi } from "vitest";
import Contact from "./Contact";
import { fireEvent, render, screen, waitFor } from "@testing-library/react";

test("renders header on Contact component", () => {
  render(<Contact />);

  const headerElement = screen.getByRole("heading");

  expect(headerElement).toBeInTheDocument();
});

test("renders two inputs with placeholdervalue name and email", () => {
  render(<Contact />);

  const nameInput = screen.getByPlaceholderText("Name");
  const emailInput = screen.getByPlaceholderText("Email");

  expect(nameInput, emailInput).toBeInTheDocument();
});

test("renders a button element", () => {
  render(<Contact />);

  const button = screen.getByTestId("test-button");

  expect(button).toBeInTheDocument();
});

test("span element is initially not vissible", () => {
  render(<Contact />);

  const span = screen.getByTestId("test-span");
  //   const span2 = screen.queryByTestId("test-span");
  //   console.log("span2 :>> ", span2);
  expect(span).not.toBeVisible();
  //   expect(span2).toBeNull();
});

test("inputs should be intially empty", () => {
  render(<Contact />);

  const nameInput = screen.getByPlaceholderText("Name");
  const emailInput = screen.getByPlaceholderText("Email");

  expect(nameInput.value).toBe("");
  expect(emailInput.value).toBe("");
});

test("inputs should display what the user types  as a controlled input", () => {
  render(<Contact />);

  const nameInput = screen.getByPlaceholderText("Name");
  const emailInput = screen.getByPlaceholderText("Email");

  const nameInputValue = "Tom";
  const emailInputValue = "tom@email.com";

  //Reproduce an event

  fireEvent.change(nameInput, { target: { value: nameInputValue } });
  fireEvent.change(emailInput, { target: { value: emailInputValue } });

  expect(nameInput.value).toBe(nameInputValue);
  expect(emailInput.value).toBe(emailInputValue);
});

test("span should display name and email after user clicks in button", async () => {
  render(<Contact />);

  const nameInput = screen.getByPlaceholderText("Name");
  const emailInput = screen.getByPlaceholderText("Email");
  const button = screen.getByTestId("test-button");
  const span = screen.getByTestId("test-span");

  const nameInputValue = "Tom";
  const emailInputValue = "tom@email.com";

  //Reproduce  events

  fireEvent.change(nameInput, { target: { value: nameInputValue } });
  fireEvent.change(emailInput, { target: { value: emailInputValue } });
  fireEvent.click(button);

  //   expect(span).toBeVisible();
  //   expect(span).toHaveTextContent(
  //     `Thank you ${nameInputValue}, we will send an email to ${emailInputValue}`
  //       );

  //? ASYNC Assert

  await waitFor(
    () => {
      expect(span).toBeVisible();
      expect(span).toHaveTextContent(
        `Thank you ${nameInputValue}, we will send an email to ${emailInputValue}`
      );
    },
    { timeout: 2000 }
  );
});

test("Contact  snapshot", () => {
  const { container } = render(<Contact />);
  console.log("container :>> ", container);
  expect(container).toMatchSnapshot();
});

test("renders userName", async () => {
  render(<Contact />);

  const userName = await screen.findByText(/Bret/i);
  expect(userName).toBeInTheDocument();
});

test("test endpoint returns userName", async () => {
  //before setting the Enviroment (render())
  const mockedResponse = {
    id: 1,
    username: "Bret",
  };

  vi.spyOn(window, "fetch").mockImplementationOnce(() => {
    return Promise.resolve({
      json: () => Promise.resolve(mockedResponse),
    });
  });
  render(<Contact />);

  const userName = await screen.findByText(/Bret/i);
  console.log("userName in test :>> ", userName);
  expect(userName).toBeInTheDocument();
});
