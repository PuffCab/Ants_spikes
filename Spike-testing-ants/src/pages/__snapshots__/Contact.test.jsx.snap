// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Contact  snapshot 1`] = `
<div>
  <div>
    <h1>
      Contact
    </h1>
    <span
      data-testid="test-span"
      style="visibility: hidden;"
    >
      Thank you 
      , we will send an email to 
    </span>
    <div>
      <input
        id="name"
        name="name"
        placeholder="Name"
        type="text"
        value=""
      />
      <input
        id="email"
        name="email"
        placeholder="Email"
        type="text"
        value=""
      />
    </div>
    <button
      data-testid="test-button"
    >
      Contact me
    </button>
  </div>
</div>
`;
